schemachanger: Implement `ADD CONSTRAINT ... CHECK`

Xiang-Gu opened the pull request:
This PR implemented `ALTER TABLE ... ADD CHECK ...` in the declarative
schema changer. It can be broken into two big categories:
 1. Enabling adding/dropping path of ConstraintName element;
 2. Implementing `ALTER TABLE ... ADD CHECK` in the builder.

Epic: None

Xiang-Gu committed:
schemachanger: enable adding/dropping path of constraint_name

ConstraintName element is deemed simple dependents of constraint and
hence has transition between PUBLIC <==> ABSENT. This commit thus
is primarily implementing the scop `SetConstraintName`.

Xiang-Gu committed:
sql: Use schema resolver during check constraint validation

During constraint validation, a side hustle is to resolve table names
and type so we can pretty print them in log/error messages. The existing
resolver is not adequate if the check constraint references sequences
or user-defined types.

Xiang-Gu committed:
schemaexpr: Refactor some existing function

`DequalifyAndValidateExpr` is used to validate an expression such as
the expr has given type, its volatility is "alright", etc. This function
will be used in the builder for `ADD CONSTRAINT ... CHECK` but it
currently expects a table descriptor, which we might not have in the
builder. To circumvent that limitation, this commit refactor this
function by plumbing in the needed column information and lookupFn.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91153)
<!-- Reviewable:end -->

Xiang-Gu committed:
scmutationexec: two small behavior changes

1. retain check constraint in the dropping path on `Checks`. The old
schema changer has the behavior of retaining the to-be-dropped
check constraint on the `Checks` slice. We will do the same in the
new schema changer.

2. when adding a check constraint, we enqueue it to the mutation slice
and then immediately fast-forward its mutation state to WRITE_ONLY.
Previously it will stay in DELTE_ONLY. This change makes it clear that
the check constraint, at this stage, is "enforced".

3. removed some dead code.

Xiang-Gu committed:
scbuildstmt: Implement `ADD CONSTRAINT ... CHECK`

This commit implements `ADD CONSTRAINT ... CHECK`. Thanks to the
previous infrastructural work for enabling adding/dropping check
constraint, we only need to add code in the new schema changer builder.

It also enabled `ADD CONSTRAINT ... CHECK` to be on by default. It
requires us to turn off new schema changer in various tests bc those
tests depends on specific behavior/knobs of the old schema changer.

chengxiong-ruan commented on pkg/sql/backfill.go:
```diff
@@ -1538,9 +1538,10 @@ func ValidateCheckConstraint(
 		ctx context.Context, txn *kv.Txn, ie sqlutil.InternalExecutor, descriptors *descs.Collection,
 	) error {
 		// Use the DistSQLTypeResolver because we need to resolve types by ID.
```
update comment?

chengxiong-ruan commented on pkg/sql/catalog/schemaexpr/expr.go:
```diff
@@ -74,29 +75,27 @@ func DequalifyAndTypeCheckExpr(
-	sourceInfo := colinfo.NewSourceInfoForSingleTable(
-		*tn, colinfo.ResultColumnsFromColumns(desc.GetID(), nonDropColumns),
-	)
+	sourceInfo := colinfo.NewSourceInfoForSingleTable(*tn, getAllNonDropColumnsFn())
```
since we always call `getAllNonDropColumnsFn()`, how about we just pass in a `colinfo.ResultColumns`?

Xiang-Gu replied:
I will wait a bit on this part bc we might want to use an interface (or something) to make this cleaner. But that will be in a different PR!

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -38,3 +75,240 @@ func alterTableAddConstraint(
+
+// alterTableAddConstraintCheck contains logic for building ALTER TABLE ... ADD CONSTRAINT ... CHECK.
+// It assumes `t` is such a command.
+func alterTableAddConstraintCheck(
```
`alterTableAddCheckConstraint`

Xiang-Gu replied:
I changed it to just `alterTableAddCheck`, in accordance to `alterTableAddPrimaryKey`.

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -38,3 +75,240 @@ func alterTableAddConstraint(
+	var nameBuf bytes.Buffer
+	nameBuf.WriteString("check")
+
+	iterateColNamesInExpr(b, tableID, expr, func(columnName string) error {
```
this is nice and I like it. but do we need to consider postgres compatibility for this? (maybe not)

Xiang-Gu replied:
I don't think so bc this is what the legacy schema changer did.

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go:
```diff
@@ -679,3 +714,37 @@ func fallBackIfVirtualColumnWithNotNullConstraint(t *tree.AlterTableAddColumn) {
 }
+
+// ExtractColumnIDsInExpr extracts column IDs used in expr. It's similar to
+// schemaexpr.ExtractColumnIDsInExpr but this function can also extract columns
```
nit: `schemaexpr.ExtractColumnIDs`

Xiang-Gu replied:
done.

chengxiong-ruan commented on pkg/sql/schemachanger/scexec/scmutationexec/references.go:
```diff
@@ -112,7 +112,11 @@ func (m *visitor) MakeAbsentCheckConstraintWriteOnly(
+	if err = enqueueAddCheckConstraintMutation(tbl, ck); err != nil {
+		return err
+	}
+	tbl.Mutations[len(tbl.Mutations)-1].State = descpb.DescriptorMutation_WRITE_ONLY
```
could you explain why we want to move the state to write_only here instead of doing it in an operation at some later point? maybe a comment would be helpful

Xiang-Gu replied:
It actually does not matter even if we don't do it (and leave it in the DELETE_ONLY state); It's good that we do this (fast-forward it to WRITE_ONLY) bc that aligns with the concept "enforced" that we introduced, that is, such a constraint, at this time, is considered "enforced" and should better have its mutation state be WRITE_ONLY.

I've added a comment here.

chengxiong-ruan reviewed (commented):
nicely done!

chengxiong-ruan commented on pkg/sql/schemachanger/scplan/internal/rules/dep_add_constraint.go:
```diff
@@ -19,16 +19,29 @@ import (
+	registerDepRule(
+		"constraint dependents exist before constraint becomes public",
+		scgraph.Precedence,
+		"dependent", "constraint",
```
should we swap `"dependent"` and `"constraint"` here?

postamar commented on pkg/sql/catalog/descpb/constraint.go:
```diff
@@ -125,3 +125,22 @@ func (c *ConstraintDetail) GetConstraintName() string {
+	case ConstraintTypeCheck:
+		c.CheckConstraint.Name = name
+	}
+}
```
As it turns out, `ConstraintDetail` is now no longer used anywhere and can be removed entirely (which I forgot to do)

Xiang-Gu replied:
yeah, I've deleted both methods (GetConstraintName and SetConstraintName) bc they are not used anywhere.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -11,16 +11,53 @@
+// panicBecauseGuardRailsIsViolated is used when previously passed guardrail, `extraChecks`, is now violated.
+func panicBecauseGuardRailsIsViolated() {
+	panic(errors.AssertionFailedf("previously passed guardrail `extraChecks` for ALTER TABLE is now violated."))
+}
```
Is this really helpful?

Xiang-Gu replied:
it's just that we will be typing this long `panic(......)` line multiple times at its call-sites. Do you have any suggestions for this situation?

postamar replied:
My point was more that we already check for this in `alter_table.go` and that checking a second time isn't helpful.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -38,3 +75,240 @@ func alterTableAddConstraint(
+	scpb.ForEachConstraintName(b.QueryByID(tbl.TableID), func(
+		_ scpb.Status, _ scpb.TargetStatus, e *scpb.ConstraintName,
+	) {
+		if e.Name == name.String() {
```
I think you're supposed to use `string(name)` here or else names which require quotes won't work.

Xiang-Gu replied:
fixed

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go:
```diff
@@ -292,6 +292,41 @@ func getSortedColumnIDsInIndex(
+	sort.Slice(ret, func(i, j int) bool {
+		return ret[i] < ret[j]
+	})
+	return ret
```
Why not return a slice of column elements here instead? We usually prefer manipulating objects directly, rather than references to objects, whenever possible.

Also I'm not sure whether you need to sort them at all. The order in which they're in should be stable, I think.

Xiang-Gu replied:
I've changed the return to `[]*scpb.Column` and got rid of the sorting step.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go:
```diff
@@ -292,6 +292,41 @@ func getSortedColumnIDsInIndex(
+	colElems := b.ResolveColumn(tableID, columnName, ResolveParams{
+		IsExistenceOptional: true,
+		RequiredPrivilege:   privilege.CREATE,
+	})
```
Using `ResolveColumn ` will not work on columns added in the same transaction:
```
BEGIN;
ALTER TABLE t ADD COLUMN c INT;
ALTER TABLE t ADD CONSTRAINT ck CHECK (c > 0);
```
While this case is not supported in prod, we should try to support it in all `pkg/sql/schemachanger/...` tests.

postamar replied:
Oh nice. I thought otherwise for some reason. Excellent! That's a relief.

postamar reviewed (commented):
Nice work! I have a few comments to make while you deal with test failures.

fqazi commented on pkg/sql/catalog/schemaexpr/expr.go:
```diff
@@ -74,29 +75,27 @@ func DequalifyAndTypeCheckExpr(
 	semaCtx *tree.SemaContext,
 	maxVolatility volatility.V,
 	tn *tree.TableName,
+	getAllNonDropColumnsFn func() colinfo.ResultColumns,
```
Would it make sense to have some type of interface, just trying to think how we can make these cleaner in general?

Xiang-Gu replied:
It definitely makes sense! I just chosen to not do it in this PR as it's always very bloated.

fqazi replied:
Its okay to leave this a as a follow on

Xiang-Gu commented:
Now that the CI is green (finally), this is ready for another look. 

Most of the combat for CI failure revolves around the improvement we made in the new schema changer that a check constraint references sequences by ID (rather than by name). That difference has consequences that "leaked" to the legacy schema changer, and I need to modify the legacy schema change to handle those consequences.

E.g., if we created a table and added a constraint that references a sequence in the new schema changer, and later drop that sequence cascade in the old schema changer, the old schema changer will see that this sequence is referenced by a table's constraint, which previously won't happen at all (a known limitation), and does not know what to do. I taught it to ignore/skip those tables.

Xiang-Gu committed:
scbuildstmt: support minClusterVersion for various `ADD CONSTRAINT` stmt

`minClusterVersion` is used to version gate each fully supported
stmts in the declarative schema changer. Previously, there is only one
`minClusterVersion` for stmt `ALTER TABLE ADD CONSTRAINT`, which is fine
bc we only supported one `ADD CONSTRAINT` (namely, `ADD PRIMARY KEY`)
so far. We will need to extend this as we are going to support more
`ADD CONSTRAINT` in the declarative schema changer.

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/index.go:
```diff
@@ -350,6 +350,29 @@ func (m *visitor) SetIndexName(ctx context.Context, op scop.SetIndexName) error
+		return err
+	}
+	if constraint.AsUniqueWithIndex() != nil {
+		constraint.AsUniqueWithIndex().IndexDesc().Name = op.Name
```
We have to be careful with index-backed constraints, because their names are already modelled by the `scpb.IndexName` element. I'm not keen on having different elements represent the same thing so can you make `ConstraintName` really represent non-index-backed-constraint names? And in this line of code, return an assertion error.

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/index.go:
```diff
@@ -350,6 +350,29 @@ func (m *visitor) SetIndexName(ctx context.Context, op scop.SetIndexName) error
+		return errors.AssertionFailedf("unknown constraint type")
+	}
+	return nil
+}
```
This op implementation should probably live in `constraint.go` instead of `index.go`.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go:
```diff
@@ -119,33 +125,53 @@ func alterTableIsSupported(n *tree.AlterTable, mode sessiondatapb.NewSchemaChang
+
+	minSupportedClusterVersion := alterTableAddConstraintMinSupportedClusterVersion[cmdKey]
+	return b.EvalCtx().Settings.Version.IsActive(b, minSupportedClusterVersion)
+}
```
I'm sorry but this code is still needlessly complex. Rather than dispatching in a switch block why not introduce a `supportedAlterTableAddConstraintStatements` map in the same vein as the other maps? That way you can get rid of all these useless `panic(errors.AssertionFailedf("previously checked guardrail violated: Unknown add constraint command type"))` which are effectively dead code. Also instead of keying on free-form strings, key on the `reflect.Type` of an AST node.

postamar reviewed (commented):
I've reviewed all but the last commits, I'll continue later. Those are coming along nicely.

postamar commented on pkg/sql/schemachanger/screl/attr.go:
```diff
@@ -172,6 +176,7 @@ var elementSchemaOptions = []rel.SchemaOption{
 		rel.EntityAttr(ConstraintID, "ConstraintID"),
 		rel.EntityAttr(ReferencedSequenceIDs, "UsesSequenceIDs"),
 		rel.EntityAttr(ReferencedTypeIDs, "UsesTypeIDs"),
+		rel.EntityAttr(IndexIDToValidate, "IndexIDToValidate"),
```
You don't actually need this new attribute type, instead `rel.EntityAttr(IndexID, "IndexIDToValidate"),` will do.

postamar commented on pkg/sql/drop_sequence.go:
```diff
@@ -227,16 +227,54 @@ func (p *planner) canRemoveOwnedSequencesImpl(
 func dropDependentOnSequence(ctx context.Context, p *planner, seqDesc *tabledesc.Mutable) error {
-	for len(seqDesc.DependedOnBy) > 0 {
-		dependent := seqDesc.DependedOnBy[0]
+	// numDependedOnByTablesToSkip is a temporary hack to skip those tables that reference
```
Instead of describing a hack as a temporary, describe the conditions which need to be fulfilled to remove the hack. One should always assume that, unless somebody's actively working on part of the code right now, it's not going to change soon.

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -329,6 +329,9 @@ message CheckConstraint {
   bool from_hash_sharded_column = 5;
+  // index_id_to_validate is the index id to hint to the check constraint validation SQL query. It is used
+  // exclusively by sql.validateCheckExpr).
+  uint32 index_id_to_validate = 6 [(gogoproto.customname) = "IndexIDToValidate", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.IndexID"];
```
can you rename this to `Index_id_for_validation` please?

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/dep_add_column_and_constraint.go:
```diff
@@ -0,0 +1,41 @@
+)
+
+// These rules ensure that columns and constraints on those columns come
+// to existence in the appropriate order.
```
As it turns out this rule binds a constraint to an index, not to columns.

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/references.go:
```diff
@@ -400,12 +401,16 @@ func updateBackReferencesInSequences(
 		return nil
 	})
 	if forwardRefs.Contains(seqID) {
+		// Add `colID` to `updated`, if not exists.
```
nit: this comment is not useful.

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/references.go:
```diff
@@ -400,12 +401,16 @@ func updateBackReferencesInSequences(
+			updated.Add(colID)
+		}
 	} else {
+		// Remove `colID` from `updated`, if exists.
```
nit: neither is this one.

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/dep_add_constraint.go:
```diff
@@ -19,15 +19,14 @@ import (
-
 	registerDepRule(
-		"constraint dependent public right before constraint",
+		"constraint dependent public right before index-backed constraint",
```
this name change is inconsistent with the definition of `isConstraint`

postamar reviewed (changes requested):
I did a pass on the last commit and didn't find anything shocking. I requested a few changes which will be noisy so I'm holding off on reviewing further. Still, I think this is close. Great work!

Xiang-Gu committed:
scpb: Rename element `ConstraintName` to `ConstraintWithoutIndexName`

We want to be rigiorous in our naming of the `ConstraintName` element,
which is really just for non-index-backed constraint.

Xiang-Gu reviewed (commented):
@postamar Thanks for your review. I've addressed your feedback and pushed the changes. If you get a chance, I'd appreciate it if you can give it another look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-:-NIikm6w5GujvVPvDU3t:bc9qgrm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, and @postamar)

___
*[`pkg/sql/backfill.go` line 1540 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NFtoMzoNVvFKaYP5fEj-r7-1540:-NIhzdf-8bUcwHSW1Uwn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01011ba41f6a2680a658d0c9074df4d7fa30596c/pkg/sql/backfill.go#L1540)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

update comment?
</blockquote></details>

Done.



_Code quote:_
```Go
rn runHistoricalTxn.Exec(ctx, func(
```
___
*[`pkg/sql/backfill.go` line 1541 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NFtoMzoNVvFKaYP5fEj-r15-1541:-NIhzfON7xKhCBxsJHNC:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/873eb5a6d3e2b56a1b2df4bc6d8f8a77edb115c7/pkg/sql/backfill.go#L1541)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: comment what the last `nil` is.
</blockquote></details>

done
___
*[`pkg/sql/drop_sequence.go` line 238 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NIEmSh8-0Ydp8EswR7q-r18-238:-NIi0uQP3lY-rFLVchWd:b5c9t98) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec336358ca0e72a1a710ba1b2e60c7a8f19a7793/pkg/sql/drop_sequence.go#L238)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Instead of describing a hack as a temporary, describe the conditions which need to be fulfilled to remove the hack. One should always assume that, unless somebody's actively working on part of the code right now, it's not going to change soon.
</blockquote></details>

done. I think my previous comments were too verbose and I've shortened it and put the condition to remove this hack in the comment as well.



_Code quote:_
```Go
umDependedOnByTablesToSki
```
___
*[`pkg/sql/catalog/descpb/constraint.go` line 108 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NFtoMzoNVvFKaYP5fEd-r13-108:-NIi1FhxBepha8OttKZe:bet65y2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9accd3f99f36a5c488432d9d0c9c1bd6ee3902ce/pkg/sql/catalog/descpb/constraint.go#L108)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Remove the `ConstraintDetail` struct definition also. It's not used at all. 
</blockquote></details>

done, thanks for pointing this out!
___
*[`pkg/sql/catalog/descpb/structured.proto` line 946 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NI9WakN8sQyyAuc9lDw-r18-946:-NIi1atN8m8xdYrdUldD:bbtdw16) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec336358ca0e72a1a710ba1b2e60c7a8f19a7793/pkg/sql/catalog/descpb/structured.proto#L946)):*
> argument
I hope our slack discussion got you on board with this solution, so I will keep it as is. If you have further doubts, let me know and we can discuss more.

For other reviewers, the most important reason that we want to add such a field in this `CheckConstraint` element is to allow writing the dep rule easily: the to-be-checked primary index needs to be ready before the check constraint reaches validated. Having such a field allows us to express the dependency rule with a join on this `index_id_to_validate` field.
___
*[`pkg/sql/catalog/schemaexpr/expr.go` line 86 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NH1yBnpIqzr5tkW2b9b-r6-86:-NIi2V024TjJi33j-TQb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90143060e105d5aae8cc6fb566204c59fd80cb98/pkg/sql/catalog/schemaexpr/expr.go#L86)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

It's okay to leave this as a follow on
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go` line 171 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NH1yBnqR_gti3ZW3yOm-r17-171:-NIi6Q29DoZSY9fKuusS:bwls9ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9ec29a84a90dc46a4d90195b50ac4b7129eb021/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L171)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm sorry but this code is still needlessly complex. Rather than dispatching in a switch block why not introduce a `supportedAlterTableAddConstraintStatements` map in the same vein as the other maps? That way you can get rid of all these useless `panic(errors.AssertionFailedf("previously checked guardrail violated: Unknown add constraint command type"))` which are effectively dead code. Also instead of keying on free-form strings, key on the `reflect.Type` of an AST node.
</blockquote></details>

A difficulty of that approach will be that we re-use the same `*tree.UniqueConstraintTableDef` type for `UNIQUE`, `UNIQUE WITHOUT INDEX`, and `PRIMARY KEY` constraints, so keying on `reflect.Type` won't work. That's why I chose to just key on free-form strings.

I can definitely remove all those panic code since we shouldn't expect those lines to be hit (I was just being paranoid). Let me know if you have a better alternative solution here (I am not very satisfied with the free-form strings either).



_Code quote:_
```Go
inSupportedClusterVersion := alterTableAddConstraintMinSupportedClusterVersion[cmdKey]
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 54 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NH1yBnqR_gti3ZW3yOn-r7-54:-NIi9Ie11l6tG2mXy9zI:b-jtp4oy) ([raw file](https://github.com/cockroachdb/cockroach/blob/01011ba41f6a2680a658d0c9074df4d7fa30596c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L54)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

My point was more that we already check for this in `alter_table.go` and that checking a second time isn't helpful.
</blockquote></details>

I removed those lines.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go` line 316 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NH1yBnqR_gti3ZW3yOo-r7-316:-NIi9p61CiL7bzihuM9B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01011ba41f6a2680a658d0c9074df4d7fa30596c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go#L316)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Oh nice. I thought otherwise for some reason. Excellent! That's a relief.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/references.go` line 404 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NH1yBnoWwK3B-vAuLo9-r18-404:-NIiDu3D4YwFixQPLo_E:bbkkphh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec336358ca0e72a1a710ba1b2e60c7a8f19a7793/pkg/sql/schemachanger/scexec/scmutationexec/references.go#L404)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: this comment is not useful.
</blockquote></details>

I spent some time understanding what this function is doing, so I added those comments. But yeah,  I think what's better is to add some comments for the whole function. To that regard, I removed those two lines of comments and added some for the whole function.



_Code quote:_
```Go
// Add `colID` to `updated`, if not exists.
```
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/references.go` line 413 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NH1yBnoWwK3B-vAuLo9-r18-413:-NIiEKHoAuF5ksxmS_sR:bwwljk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec336358ca0e72a1a710ba1b2e60c7a8f19a7793/pkg/sql/schemachanger/scexec/scmutationexec/references.go#L413)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: neither is this one.
</blockquote></details>

acked above
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 334 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NI9WakOF-ayDPLt-gcc-r18-334:-NIiEuxzEKQh-nu2DHFx:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec336358ca0e72a1a710ba1b2e60c7a8f19a7793/pkg/sql/schemachanger/scpb/elements.proto#L334)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

can you rename this to `Index_id_for_validation` please?
</blockquote></details>

done



_Code quote:_
```protobuf
 index_id_to_validate = 6 [(gogoproto.cus
```
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_add_constraint.go` line 23 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NI7vREok7DrBDgcdfIE-r18-23:-NIiFTywDBlpNum1CN7L:b1rkcms) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec336358ca0e72a1a710ba1b2e60c7a8f19a7793/pkg/sql/schemachanger/scplan/internal/rules/dep_add_constraint.go#L23)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

this name change is inconsistent with the definition of `isConstraint`
</blockquote></details>

thanks for catching this. Fixed
___
*[`pkg/sql/schemachanger/screl/attr.go` line 179 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NI9WakOF-ayDPLt-gci-r18-179:-NIiGCVj8qKwjL5TF5Ep:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec336358ca0e72a1a710ba1b2e60c7a8f19a7793/pkg/sql/schemachanger/screl/attr.go#L179)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You don't actually need this new attribute type, instead `rel.EntityAttr(IndexID, "IndexIDToValidate"),` will do.
</blockquote></details>

done



_Code quote:_
```Go
rel.EntityMapping(t((*scpb.CheckConstraint)(nil)),
```
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/index.go` line 363 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NFtoMzoNVvFKaYP5fEe-r14-363:-NIifJvo3Q0TAv0flY_x:b1g2jso) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ee9dd5c12ebedf74c0d3abe0cd3554944397c14/pkg/sql/schemachanger/scexec/scmutationexec/index.go#L363)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

We have to be careful with index-backed constraints, because their names are already modelled by the `scpb.IndexName` element. I'm not keen on having different elements represent the same thing so can you make `ConstraintName` really represent non-index-backed-constraint names? And in this line of code, return an assertion error.
</blockquote></details>

Done. I renamed the element to `ConstraintWithoutIndexName`.
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/index.go` line 374 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NFtoMzoNVvFKaYP5fEe-r14-374:-NIiBwWG6ArThYREbeVX:b-w8wrj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ee9dd5c12ebedf74c0d3abe0cd3554944397c14/pkg/sql/schemachanger/scexec/scmutationexec/index.go#L374)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This op implementation should probably live in `constraint.go` instead of `index.go`.
</blockquote></details>

agreed. I created a new `constraint.go` in this package and moved all the op implementations related to constraints there.



_Code quote:_
```Go
	} else if constraint.AsUniqueWithoutIndex() != nil {
		constraint.AsUniqueWithoutIndex().UniqueWithoutIndexDesc().Name = op.Name
	} else if constraint.AsCheck() != nil {
		constraint.AsCheck().CheckDesc().Name = op.Name
```
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_add_column_and_constraint.go` line 21 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/91153#-NI9WakOF-ayDPLt-gck-r18-21:-NIiFCf0AF7lduCwiO8M:b-ngpp4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec336358ca0e72a1a710ba1b2e60c7a8f19a7793/pkg/sql/schemachanger/scplan/internal/rules/dep_add_column_and_constraint.go#L21)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

As it turns out this rule binds a constraint to an index, not to columns.
</blockquote></details>

done. I've changed the comments as well as the file name.



_Code quote:_
```Go
// These rules ensure that columns and constraints on those columns come
// to existence in the appropriate order.
```


<!-- Sent from Reviewable.io -->

postamar commented:
FYI, you forgot about removing `index_id_to_validate` from `structured.proto` which was arguably my number 1 concern.

Xiang-Gu commented:
> FYI, you forgot about removing `index_id_to_validate` from `structured.proto` which was arguably my number 1 concern.

Don't we need it? Otherwise, how could we pass this information from scpb.CheckConstraint to a descpb.TableDescriptor_CheckConstraint?

(I indeed forget to change its name to index_id_for_validation though)

Xiang-Gu commented:
@postamar Sorry, I didn't fully understand you previously. Yeah, I agree it's nicer to leave `structure.proto` alone and plumb in the index id for validation via `scop.ValidateCheckConstraint` operation. I've pushed the changes and can you retake a look at it? Thank you so much!

postamar reviewed (approved):
Excellent, thanks! We really don't want to touch those `descpb` protobufs unless we need to.

Xiang-Gu commented:
TFTR!

bors r+
